### PR TITLE
feat(qontract-api): OPA-based authorization

### PR DESCRIPTION
**Pull request by chassing**
## Summary

- Enables role-based access control [APPSRE-14305](https://redhat.atlassian.net/browse/APPSRE-14305)
- Add OPA (Open Policy Agent) authorization to qontract-api, preventing arbitrary Vault secret reads and credential exfiltration via SSRF ([APPSRE-14303](https://issues.redhat.com/browse/APPSRE-14303))
- OPA sidecar validates every request's parameters (Vault paths, service URLs) against role-based policies with regex matching
- Design doc: https://github.com/openshift-online/platform-engineering-enhancements/blob/main/app-sre/design-docs/qontract-api-opa-authorization.md

## App-Interface Dependency

https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/191745

### Milestone 1: OPA client + Rego policies

- `qontract_api/qontract_api/opa.py` — `OPAClient` (fail-closed, connection pooling, audit logging, Prometheus metrics) + `flatten_params()` for POST body dot-notation
- `qontract_api/qontract_api/dependencies.py` — authZ integrated into `UserDep` (`_authenticate` → `_authorize` → `get_current_user`), zero endpoint changes needed
- `qontract_api/qontract_api/health.py` — OPA health in `/health/ready` readiness probe
- `qontract_api/qontract_api/config.py` — `OPASettings` with `enabled`, `host`, `timeout`, `skip_endpoints`
- `opa/authz/` — `rbac.rego` + `user.rego` (reused from automated-actions)
- `opa/dev/roles.yml` — documented dev policies with SSRF-prevention examples
- `opa/Dockerfile` — UBI-based OPA image with Rego tests in build
- 19 Python tests, 20 Rego tests (SSRF prevention, Vault path restriction, token blocking)
- `docs/adr/ADR-021-opa-authorization.md`

### Milestone 2: Production deployment

- `.tekton/opa-master-push.yaml` + `opa-master-pull-request.yaml` — Konflux build pipelines
- `openshift/qontract-api.yaml` — OPA sidecar in API Deployment with probes, `QAPI_OPA_ENABLED` toggle
- `generate_token.py` — appends `-YYYY-MM-DD` for unique subjects (per-token blocking via OPA)

## Behavioral changes

- **Readiness probe status**: Before this PR, an unhealthy component caused `/health/ready` to report `"degraded"`; now it correctly reports `"unhealthy"`. This is more semantically correct (degraded implies partial functionality, unhealthy means a required component is down). Operators should be aware that alerting thresholds based on the `"degraded"` string will need updating.

## Test plan

- [x] `cd qontract_api && make test` — 452 passed
- [x] `cd opa && make test` — Rego tests pass (SSRF prevention, token blocking)
- [x] `uv run mypy` — clean
- [x] `uv run ruff check` — clean
- [ ] Konflux picks up OPA component and builds image
- [ ] Stage deployment with OPA sidecar + dev policy
- [ ] Verify: authorized request → 200, unauthorized params → 403, health endpoint → bypasses OPA

## Summary by CodeRabbit

* **New Features**
  * Optional Open Policy Agent (OPA) sidecar for API authorization with role-based, parameter-aware access controls and fail-closed enforcement
  * Added OPA decision metrics and expanded health checks to include OPA component status
  * JWT token subject generation now includes today’s UTC date stamp
* **Deployment**
  * OpenShift template updated to enable/configure the OPA sidecar and resource sizing
  * CI pipelines added to build/publish OPA images on pull requests and pushes
* **Documentation**
  * ADR added describing the OPA authorization design and threat mitigations
* **Tests**
  * Expanded OPA policy/client test coverage for parameter edge cases, skip-routes, and authorization outcomes

### CODE REVIEW

File: opa/authz/rbac.rego [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	# Check that null values in expected mean that the key should not be present in provided
+	null_keys := {k | expected[k] == null}
+	every k in null_keys {
+		not provided[k]
+	}
+
+	# For non-null values, ensure they match using regex
+	non_null_keys := {k | expected[k] != null}
+	every k in non_null_keys {
+		regex.match(sprintf("(?i)%s", [expected[k]]), provided[k])
```
[the remark this answers is not in the document]
Comment (chassing) [reply opa/authz/rbac.rego:46]: Intentional — policy values are regex patterns by design. The policy data comes from a reviewed ConfigMap in app-interface, not from user input. Identical pattern to automated-actions.

File: opa/authz/user_test.rego [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+test_unknown_user_gets_default_objs_only if {
+	objects := authz.objects with input as {"username": "unknown-user"}
+		with data.users as mock_users
+		with data.roles as mock_roles
+	"liveness" in objects
+	not "ldap-users-check" in objects
+	not "*" in objects
+}
```
[the remark this answers is not in the document]
Comment (chassing) [reply opa/authz/user_test.rego:49]: Added edge-case tests in `rbac_test.rego`: null params (must-not-be-present), empty params, and missing required params. Fixed in 6f791e1d.

File: opa/dev/roles.yml [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+# preventing VCS tokens from being sent to attacker-controlled repos.
+#
+# PARAMETER MATCHING
+# ==================
+#
+# - All param values are case-insensitive regex patterns
+# - Parameters not listed in a role permission are not checked
+#   (extra params in the request are allowed)
+# - A param set to null means it must NOT be present in the request
+
```
[the remark this answers is not in the document]
Comment (chassing) [reply opa/dev/roles.yml:31]: Regex patterns are intentional — this is the core mechanism for parameter matching. The values are reviewed via app-interface MRs, not user-supplied.

File: opa/dev/roles.yml [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+roles:
+  # ── SSRF prevention: LDAP ──────────────────────────────────────
+  # Restricts which LDAP servers and Vault paths can be used.
+  # Prevents: attacker provides ldap://evil.com + arbitrary Vault path
+  ldap-role:
+    - obj: ldap-users-check
+      params:
+        secret.server_url: "^ldaps?://freeipa\\.example\\.com$"
+        secret.path: "^app-sre/creds/ldap"
+        secret.secret_manager_url: "^https://vault\\.example\\.com$"
```
[the remark this answers is not in the document]
Comment (chassing) [reply opa/dev/roles.yml:67]: Fixed in 6f791e1d — added end anchors to all path patterns.

File: opa/dev/roles.yml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        path: "^app-sre/creds/"
+    - obj: vcs-get-file
+      params:
+        repo_url: "^https://(github\\.com|gitlab\\.cee\\.redhat\\.com)/"
+        path: "^app-sre/creds/"
+    - obj: vcs-file-sync
+      params:
+        token.path: "^app-sre/creds/"
+        repo_url: "^https://(github\\.com|gitlab\\.cee\\.redhat\\.com)/"
+
```
[the remark this answers is not in the document]
Comment (chassing) [reply opa/dev/roles.yml:85]: Fixed in 6f791e1d.

File: opa/dev/roles.yml [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  # ── Secret-read prevention: PagerDuty ──────────────────────────
+  slack-pd-role:
+    - obj: pagerduty-schedule-users
+      params:
+        path: "^app-sre/creds/pagerduty"
+    - obj: pagerduty-escalation-policy-users
+      params:
+        path: "^app-sre/creds/pagerduty"
+
```
[the remark this answers is not in the document]
Comment (chassing) [reply opa/dev/roles.yml:94]: Fixed in 6f791e1d.

File: opa/dev/roles.yml [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  # ── Secret-read prevention: Slack ──────────────────────────────
+  slack-chat-role:
+    - obj: slack-chat-post-message
+      params:
+        secret.path: "^app-sre/creds/slack"
+
```
[the remark this answers is not in the document]
Comment (chassing) [reply opa/dev/roles.yml:100]: Fixed in 6f791e1d.

File: opa/dev/roles.yml [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  slack-vcs-role:
+    - obj: vcs-repo-owners
+      params:
+        repo_url: "^https://(github\\.com|gitlab\\.cee\\.redhat\\.com)/"
+        path: "^app-sre/creds/"
+    - obj: vcs-get-file
+      params:
+        repo_url: "^https://(github\\.com|gitlab\\.cee\\.redhat\\.com)/"
+        path: "^app-sre/creds/"
+
```
[the remark this answers is not in the document]
Comment (chassing) [reply opa/dev/roles.yml:111]: Fixed in 6f791e1d.

File: qontract_api/qontract_api/dependencies.py [hunk 9]
```diff
@@ -40,6 +39,45 @@ def get_current_user(
...
+        return
+
+    params: dict[str, str] = {}
+    params.update({k: str(v) for k, v in request.path_params.items()})
+    params.update({k: str(v) for k, v in request.query_params.items()})
+
+    with contextlib.suppress(Exception):
+        body = await request.json()
+        if isinstance(body, dict):
+            params.update(flatten_params(body))
```
[the remark this answers is not in the document]
Comment (chassing) [reply qontract_api/qontract_api/dependencies.py:58]: Our endpoints don't use multi-value query params — all query params are single-valued (Pydantic models enforce this). Not applicable here.

File: qontract_api/qontract_api/health.py [hunk 10]
```diff
@@ -60,11 +68,36 @@ def check_cache_health(cache: CacheBackend | None) -> HealthStatus:
...
+            return HealthStatus(status="healthy", message="OPA sidecar reachable")
+        return HealthStatus(
+            status="unhealthy",
+            message=f"OPA returned {response.status_code}",
+        )
+    except OSError as e:
+        return HealthStatus(
+            status="unhealthy",
+            message=f"OPA health check failed: {e}",
+        )
```
[the remark this answers is not in the document]
Comment (chassing) [reply qontract_api/qontract_api/health.py:88]: `get_health_status` is a sync function (FastAPI runs it in a threadpool). Making `check_opa_health` async would require making the entire health dependency chain async. The health probe runs every 60s — sync is acceptable here.

File: qontract_api/tests/conftest.py [hunk 11]
```diff
@@ -22,6 +22,8 @@ def pytest_configure() -> None:
     os.environ.setdefault(
         "QAPI_SECRETS__PROVIDERS", '[{"url": "https://vault.example.org"}]'
     )
+    # Disable OPA authorization for tests (no OPA sidecar available)
+    os.environ.setdefault("QAPI_OPA__ENABLED", "false")
```
[the remark this answers is not in the document]
Comment (chassing) [reply qontract_api/tests/conftest.py]: Fixed in 6f791e1d — changed to hard assignment.

File: qontract_api/qontract_api/health.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (chassing) [reply qontract_api/qontract_api/health.py:88]: Good catch — fixed in 5e7b54c8. Added `httpx.TimeoutException` to the except clause.

File: qontract_api/qontract_api/health.py [hunk 12]
```diff
@@ -60,11 +68,36 @@ def check_cache_health(cache: CacheBackend | None) -> HealthStatus:
...
 
-def get_health_status(cache: CacheOptionalDep) -> HealthResponse:
+def check_opa_health(opa_client: OPAClient | None) -> HealthStatus:
+    """Check OPA sidecar health."""
+    if opa_client is None:
+        return HealthStatus(status="healthy", message="OPA authorization disabled")
+
+    opa_health_url = opa_client.opa_url.rsplit("/v1/", maxsplit=1)[0] + "/health"
+    try:
+        response = httpx.get(opa_health_url, timeout=2)
```
Comment (hemslo) [inline qontract_api/qontract_api/health.py]: Sync HTTP call in an async context — `httpx.get()` blocks the event loop for the full 2s timeout. Should use the async OPA client or at minimum an `asyncio.get_event_loop().run_in_executor()` call. Risk is low here (health probe, short timeout) but it's inconsistent with the rest of the async design.

File: qontract_api/qontract_api/health.py [hunk 13]
```diff
@@ -60,11 +68,36 @@ def check_cache_health(cache: CacheBackend | None) -> HealthStatus:
         return HealthStatus(status="unhealthy", message=f"Cache check failed: {e}")
 
 
-def get_health_status(cache: CacheOptionalDep) -> HealthResponse:
+def check_opa_health(opa_client: OPAClient | None) -> HealthStatus:
+    """Check OPA sidecar health."""
+    if opa_client is None:
+        return HealthStatus(status="healthy", message="OPA authorization disabled")
+
+    opa_health_url = opa_client.opa_url.rsplit("/v1/", maxsplit=1)[0] + "/health"
```
Comment (hemslo) [inline qontract_api/qontract_api/health.py]: Fragile URL reconstruction: if `opa_url` doesn't contain `/v1/`, `rsplit` returns the original string unchanged and the concatenated health URL will be wrong. Consider storing the OPA base host separately in `OPAClient` or `OPASettings` instead of re-deriving it here.

File: qontract_api/qontract_api/health.py [hunk 14]
```diff
@@ -74,14 +107,17 @@ def get_health_status(cache: CacheOptionalDep) -> HealthResponse:
...
     components["cache"] = check_cache_health(cache)
 
+    # Check OPA
+    components["opa"] = check_opa_health(opa_client)
+
     # Determine overall status
     component_statuses = [c.status for c in components.values()]
     if all(s == "healthy" for s in component_statuses):
         overall_status = "healthy"
     elif any(s == "unhealthy" for s in component_statuses):
```
Comment (hemslo) [inline qontract_api/qontract_api/health.py:116]: This silently swaps the readiness probe behavior: before this PR, an unhealthy component caused `/health/ready` to report `"degraded"`; now it reports `"unhealthy"`. The new logic is more semantically correct, but it's a behavioral change that can affect alerting thresholds and rollout probes. Worth calling out in the PR description so operators know to expect the change.

File: qontract_api/qontract_api/dependencies.py [hunk 15]
```diff
@@ -40,6 +39,47 @@ def get_current_user(
...
+        return
+
+    if opa_client.should_skip(request.url.path):
+        return
+
+    params: dict[str, str] = {}
+    params.update({k: str(v) for k, v in request.path_params.items()})
+    params.update({k: str(v) for k, v in request.query_params.items()})
+
+    with contextlib.suppress(Exception):
```
Comment (hemslo) [inline qontract_api/qontract_api/dependencies.py]: `contextlib.suppress(Exception)` is very broad — it silences everything including `asyncio.CancelledError`, memory errors, and unexpected bugs. Consider narrowing to `(ValueError, UnicodeDecodeError)` or catching the specific JSON decode exception from the http library, so real errors surface.

File: qontract_api/qontract_api/scripts/generate_token.py [hunk 16]
```diff
@@ -48,12 +47,15 @@ def main() -> None:
 
     args = parser.parse_args()
 
-    token_data = TokenData(sub=args.subject)
+    today = datetime.now(tz=UTC).strftime("%Y-%m-%d")
+    subject = f"{args.subject}-{today}"
```
Comment (hemslo) [inline qontract_api/qontract_api/scripts/generate_token.py:51]: Auto-appending the date to the subject is a breaking change for any existing automation or scripts that pass `--subject` and expect the exact value to appear in the JWT. Any existing tokens also won't match OPA policy `users` entries until the ConfigMap is updated simultaneously. Consider adding a `--no-date-suffix` flag to keep backward compatibility, and document the migration path.

File: .tekton/opa-master-pull-request.yaml [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+  - name: output-image
+    value: quay.io/redhat-user-workloads/app-sre-tenant/qontract-reconcile-master/opa-master:on-pr-{{revision}}
+  - name: image-expires-after
+    value: 5d
+  - name: dockerfile
+    value: opa/Dockerfile
+  - name: path-context
+    value: .
+  - name: target-stage
+    value: prod
```
Comment (hemslo) [inline .tekton/opa-master-pull-request.yaml:33]: The `COPY --from=test ${IS_TESTED_FLAG}` cross-stage dependency in the Dockerfile should force Buildah to run the test stage, but it's safer to make it explicit: add `- name: skip-unused-stages` / `  value: "false"` after `target-stage`. Other pipelines in this repo (e.g. `qontract-api-master-pull-request.yaml`) don't use this yet either — but it's the belt-and-suspenders approach the shared pipeline supports.

File: qontract_api/qontract_api/health.py (hunk 13 above)
Comment (hemslo) [inline qontract_api/qontract_api/health.py]: Fragile URL reconstruction: if `opa_url` doesn't contain `/v1/`, `rsplit` returns the original string unchanged and the concatenated health URL will be wrong. Consider storing the OPA base host separately in `OPAClient` or `OPASettings` instead of re-deriving it here.

File: qontract_api/qontract_api/health.py (hunk 12 above)
Comment (hemslo) [inline qontract_api/qontract_api/health.py]: Sync HTTP call in an async context — `httpx.get()` blocks the event loop for the full 2s timeout. Risk is low (health probe, short timeout) but inconsistent with the async OPA client used everywhere else. Consider using `asyncio.get_event_loop().run_in_executor()` or storing the async client on app state for reuse.

File: qontract_api/qontract_api/health.py (hunk 14 above)
Comment (hemslo) [inline qontract_api/qontract_api/health.py:116]: This silently swaps the readiness probe behavior: before this PR, an unhealthy component caused `/health/ready` to report `"degraded"`; after, it reports `"unhealthy"`. The new logic is more semantically correct, but it's a behavioral change that can affect alerting thresholds and rollout probes. Worth calling out in the PR description so operators know to expect it.

File: qontract_api/qontract_api/dependencies.py (hunk 15 above)
Comment (hemslo) [inline qontract_api/qontract_api/dependencies.py]: `contextlib.suppress(Exception)` is too broad — it silences `asyncio.CancelledError`, memory errors, and unexpected bugs. Consider narrowing to `(ValueError, UnicodeDecodeError)` or the specific JSON decode exception, so real errors surface.

File: qontract_api/qontract_api/scripts/generate_token.py (hunk 16 above)
Comment (hemslo) [inline qontract_api/qontract_api/scripts/generate_token.py:51]: Auto-appending the date breaks backward compatibility: existing scripts that pass `--subject qontract-reconcile` will get `qontract-reconcile-2026-06-10` instead, which won't match existing OPA policy `users` entries or JWT revocation lists until the ConfigMap is updated simultaneously. Consider a `--no-date-suffix` flag, and document the migration path for existing tokens.

File: .tekton/opa-master-pull-request.yaml (hunk 17 above)
Comment (hemslo) [inline .tekton/opa-master-pull-request.yaml:33]: The `COPY --from=test ${IS_TESTED_FLAG}` cross-stage dependency in the Dockerfile should force Buildah to run the test stage, but it's safer to make it explicit. Add `- name: skip-unused-stages` / `  value: "false"` after `target-stage: prod` — the shared pipeline supports this param and it makes the intent clear regardless of how Buildah resolves cross-stage references.

File: qontract_api/qontract_api/health.py (hunk 13 above)
Comment (chassing) [reply qontract_api/qontract_api/health.py]: Fixed in 5493e574 — `OPAClient` now takes `host` + `package_name` and builds URLs internally, exposing a `health_url` property. No more `rsplit` URL parsing.

Comment (chassing) [reply qontract_api/qontract_api/health.py]: Fixed in 5493e574 — `OPAClient` now takes `host` + `package_name` and builds URLs internally, exposing a `health_url` property.

File: qontract_api/qontract_api/health.py (hunk 12 above)
Comment (chassing) [reply qontract_api/qontract_api/health.py]: `get_health_status` is a sync dependency — FastAPI runs it in a thread pool, so the blocking `httpx.get()` doesn't affect the event loop. We can't make it async because `check_cache_health` and the Redis client are also sync.

Comment (chassing) [reply qontract_api/qontract_api/health.py]: `get_health_status` is sync (FastAPI runs it in a thread pool). Can't convert to async because `check_cache_health` and the Redis client are also sync.

File: qontract_api/qontract_api/health.py (hunk 14 above)
Comment (chassing) [reply qontract_api/qontract_api/health.py:116]: Good call — added a "Behavioral changes" section to the PR description noting the `degraded` → `unhealthy` change for operators.

Comment (chassing) [reply qontract_api/qontract_api/health.py:116]: Added to the PR description under "Behavioral changes".

File: qontract_api/qontract_api/dependencies.py (hunk 15 above)
Comment (chassing) [reply qontract_api/qontract_api/dependencies.py]: Fixed in 5493e574 — narrowed to `contextlib.suppress(ValueError, UnicodeDecodeError)`.

Comment (chassing) [reply qontract_api/qontract_api/dependencies.py]: Fixed in 5493e574 — narrowed to `contextlib.suppress(ValueError, UnicodeDecodeError)`.

File: qontract_api/qontract_api/scripts/generate_token.py (hunk 16 above)
Comment (chassing) [reply qontract_api/qontract_api/scripts/generate_token.py:51]: This script doesn't have existing callers or backward compat concern. The date suffix is intentional design for per-token blocking via OPA. The script is temporary and will be replaced by a proper management CLI.

Comment (chassing) [reply qontract_api/qontract_api/scripts/generate_token.py:51]: No backward compat concern — this script is new in this PR and temporary (will be replaced by proper JWT management CLI). The date suffix is intentional for OPA per-token blocking.

File: .tekton/opa-master-pull-request.yaml (hunk 17 above)
Comment (chassing) [reply .tekton/opa-master-pull-request.yaml:33]: Keeping as-is — `COPY --from=test` already forces the test stage. We don't use `skip-unused-stages` anywhere in the repo, and tests should also run in the local dev environment (not only via pipeline params).

Comment (chassing) [reply .tekton/opa-master-pull-request.yaml:33]: Keeping as-is — `COPY --from=test` forces the test stage. We don't use `skip-unused-stages` elsewhere in the repo.

File: qontract_api/qontract_api/health.py (hunk 12 above)
Comment (hemslo) [reply qontract_api/qontract_api/health.py]: Should we create both sync and async redis clients? Then we can use async one in FastAPI, sync one in sync tasks, currently health check can fail when load is high.

Comment (chassing) [reply qontract_api/qontract_api/health.py]: TBH, I would like to get rid of celery and replace it with an async framework, but `qontract-utils` is and will be used by `reconcile,` and therefore we need to "stick" with the sync world as well. 
Yes, supporting both async and sync would make sense, but it would double the burden.
Definitely, it would be overkill for this PR.

**Review by hemslo — APPROVED** [review]
LGTM!
